### PR TITLE
demo-test-2

### PR DESCRIPTION
**Pull request by hilladalins**

### CODE REVIEW

File: api_views/books.py [hunk 1]
```diff
@@ -22,30 +22,43 @@ def add_new_book():
...
+            user_id=user.id
+        )
+        db.session.add(new_book)
+        db.session.commit()
+        responseObject = {
+            'status': 'success',
+            'message': 'Book has been added.'
+        }
+        return Response(json.dumps(responseObject), 201, mimetype="application/json")
+    except Exception as e:
```
Comment (hilladalins) [inline api_views/books.py:59]: Use more specific exception handling instead of catching generic Exception. Consider catching SQLAlchemyError for database operations.

File: api_views/users.py [hunk 2]
```diff
@@ -206,3 +206,64 @@ def delete_user(username):
                 return Response(error_message_helper("User not found!"), 404, mimetype="application/json")
         else:
             return Response(error_message_helper("Only Admins may delete users!"), 401, mimetype="application/json")
+
+
+def get_user_by_email(email):
```
Comment (hilladalins) [inline api_views/users.py:211]: Missing unit tests for the get_user_by_email function. Add tests to verify email lookup functionality.

File: api_views/users.py [hunk 3]
```diff
@@ -206,3 +206,64 @@ def delete_user(username):
...
+        return Response(error_message_helper(resp), 401, mimetype="application/json")
+
+    current_user = User.query.filter_by(username=resp).first()
+    if not current_user.admin:
+        return Response(error_message_helper("Only Admins may promote users!"), 401, mimetype="application/json")
+
+    user = User.query.filter_by(username=username).first()
+    if user:
+        user.admin = True
+        db.session.commit()
```
Comment (hilladalins) [inline api_views/users.py:237]: Add error handling for database commit operation to handle potential SQLAlchemy exceptions.

File: api_views/users.py [hunk 4]
```diff
@@ -206,3 +206,64 @@ def delete_user(username):
...
+    if "expired" in resp or "Invalid token" in resp:
+        return Response(error_message_helper(resp), 401, mimetype="application/json")
+
+    current_user = User.query.filter_by(username=resp).first()
+    if not current_user.admin:
+        return Response(error_message_helper("Only Admins may deactivate user accounts!"), 401, mimetype="application/json")
+
+    user = User.query.filter_by(username=username).first()
+    if user:
+        user.active = False  # Assuming 'active' is a column to manage account status
```
Comment (hilladalins) [inline api_views/users.py:261]: Add a comment explaining the implications of deactivating a user account and what 'active' status means.

File: api_views/users.py [hunk 5]
```diff
@@ -206,3 +206,64 @@ def delete_user(username):
...
+    Retrieve user details by email address.
+    """
+    user = User.query.filter_by(email=email).first()
+    if user:
+        return jsonify(user.json()), 200
+    else:
+        return Response(error_message_helper(), 404, mimetype="application/json")
+
+
+def promote_user_to_admin(username):
```
Comment (hilladalins) [inline api_views/users.py:222]: Missing unit tests for the promote_user_to_admin function. Add tests to verify admin promotion functionality.

File: api_views/users.py (hunk 2 above)
Comment (hilladalins) [inline api_views/users.py:211]: The user retrieval functions lack unit test coverage for edge cases like malformed emails or special characters in usernames

File: api_views/users.py [hunk 6]
```diff
@@ -206,3 +206,64 @@ def delete_user(username):
...
+    """
+    Promote a user to admin status. Only accessible by current admins.
+    """
+    resp = token_validator(request.headers.get('Authorization'))
```
Comment (hilladalins) [inline api_views/users.py:226]: Token validation checks are inconsistently handled across different admin functions - some use string matching while others use explicit status checks

File: api_views/users.py (hunk 4 above)
Comment (hilladalins) [inline api_views/users.py:261]: Missing validation on account deactivation to prevent an admin from deactivating their own account or other admin accounts

File: api_views/books.py [hunk 7]
```diff
@@ -22,30 +22,43 @@ def add_new_book():
...
-                'message': 'Book has been added.'
-            }
-            return Response(json.dumps(responseObject), 200, mimetype="application/json")
+    # Check if the user already has this book title
+    existing_book = Book.query.filter_by(user=user, book_title=request_data.get('book_title')).first()
+    if existing_book:
+        return Response(error_message_helper("Book already exists!"), 400, mimetype="application/json")
+    
+    # Add the new book
+    try:
```
Comment (hilladalins) [inline api_views/books.py:46]: Generic try-except block masks specific database errors that should be handled separately for proper error reporting

File: api_views/books.py [hunk 8]
```diff
@@ -22,30 +22,43 @@ def add_new_book():
...
-                           user_id=user.id)
-            db.session.add(newBook)
-            db.session.commit()
-            responseObject = {
-                'status': 'success',
-                'message': 'Book has been added.'
-            }
-            return Response(json.dumps(responseObject), 200, mimetype="application/json")
+    # Check if the user already has this book title
+    existing_book = Book.query.filter_by(user=user, book_title=request_data.get('book_title')).first()
```
Comment (hilladalins) [inline api_views/books.py:41]: Book title uniqueness check lacks case sensitivity consideration and normalization

File: api_views/users.py [hunk 9]
```diff
@@ -206,3 +206,64 @@ def delete_user(username):
...
+
+def get_user_by_email(email):
+    """
+    Retrieve user details by email address.
+    """
+    user = User.query.filter_by(email=email).first()
+    if user:
+        return jsonify(user.json()), 200
+    else:
+        return Response(error_message_helper(), 404, mimetype="application/json")
```
Comment (hilladalins) [inline api_views/users.py:219]: The error_message_helper() function is called without a message parameter, which could lead to an empty or malformed error response. Please provide an appropriate error message.

File: api_views/users.py [hunk 10]
```diff
@@ -206,3 +206,64 @@ def delete_user(username):
...
+
+
+def promote_user_to_admin(username):
+    """
+    Promote a user to admin status. Only accessible by current admins.
+    """
+    resp = token_validator(request.headers.get('Authorization'))
+    if "expired" in resp or "Invalid token" in resp:
```
Comment (hilladalins) [inline api_views/users.py:227]: The condition 'Invalid token in resp' should be checked using string comparison with equality operator (==) rather than using 'in' operator for more precise matching.

File: api_views/users.py (hunk 4 above)
Comment (hilladalins) [inline api_views/users.py:261]: Consider adding a comment explaining the implications of deactivating a user account and whether this is reversible. Also, verify that the 'active' column exists in the User model.

File: api_views/users.py (hunk 2 above)
Comment (hilladalins) [inline api_views/users.py:211]: New function get_user_by_email() appears to lack unit tests. Please add corresponding test cases to verify functionality.

File: api_views/users.py (hunk 5 above)
Comment (hilladalins) [inline api_views/users.py:222]: New function promote_user_to_admin() requires unit tests to verify admin authorization and promotion logic.

File: api_views/users.py [hunk 11]
```diff
@@ -206,3 +206,64 @@ def delete_user(username):
...
+        responseObject = {
+            'status': 'success',
+            'message': 'User promoted to admin.'
+        }
+        return Response(json.dumps(responseObject), 200, mimetype="application/json")
+    else:
+        return Response(error_message_helper("User not found!"), 404, mimetype="application/json")
+
+
+def deactivate_user_account(username):
```
Comment (hilladalins) [inline api_views/users.py:247]: New function deactivate_user_account() needs unit tests to ensure proper account deactivation and admin authorization checks.

File: api_views/books.py (hunk 8 above)
Comment (hilladalins) [inline api_views/books.py:41]: Consider using a case-insensitive comparison for book titles to prevent duplicate entries with different capitalizations.

File: api_views/books.py (hunk 1 above)
Comment (hilladalins) [inline api_views/books.py:59]: Consider using a more specific exception type instead of catching generic Exception. This broad exception handling could mask important errors. Consider handling specific exceptions like SQLAlchemyError for database operations.

File: api_views/users.py (hunk 9 above)
Comment (hilladalins) [inline api_views/users.py:219]: error_message_helper() is called without any message parameter. This will result in an empty error message. Should include a descriptive message like 'User not found'.

File: api_views/users.py (hunk 10 above)
Comment (hilladalins) [inline api_views/users.py:227]: The condition 'Invalid token in resp' relies on string matching which is fragile. Consider using specific error codes or enums instead for more robust error handling.

File: api_views/users.py (hunk 4 above)
Comment (hilladalins) [inline api_views/users.py:261]: The comment about 'active' column assumption should be verified. If this column doesn't exist, it will cause a runtime error. Consider adding a database migration check.

File: api_views/users.py (hunk 2 above)
Comment (hilladalins) [inline api_views/users.py:211]: New function get_user_by_email() was added but no corresponding unit tests were found in the changeset. Please add unit tests to verify the functionality.

File: api_views/users.py (hunk 5 above)
Comment (hilladalins) [inline api_views/users.py:222]: New function promote_user_to_admin() was added but no corresponding unit tests were found in the changeset. Please add unit tests to verify the functionality.

File: api_views/users.py (hunk 11 above)
Comment (hilladalins) [inline api_views/users.py:247]: New function deactivate_user_account() was added but no corresponding unit tests were found in the changeset. Please add unit tests to verify the functionality.
